clippy_lints/src/else_if_without_else.rs

```diff
@@ -49,24 +49,22 @@ declare_clippy_lint! {
 declare_lint_pass!(ElseIfWithoutElse => [ELSE_IF_WITHOUT_ELSE]);
 
 impl EarlyLintPass for ElseIfWithoutElse {
-    fn check_expr(&mut self, cx: &EarlyContext<'_>, mut item: &Expr) {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, item: &Expr) {
         if in_external_macro(cx.sess(), item.span) {
             return;
         }
 
-        while let ExprKind::If(_, _, Some(ref els)) = item.kind {
```

Comment:
I'm guessing this is to lint nested blocks (albeit weirdly?). Can you help to also add a test case for this?

Reply:
This caused the lint to emit on any `if-else-if-else...` chain that did not end with an `else`.  The lint would flag at the top of the chain, and then also flag each time `EarlyLintPass::check_expr` was called on the subsequent expressions in the chain. 

I added a test with 4 chains:
```
if bla1() {
     println!("if");
} else if bla2() {
     println!("else if 1");
} else if bla3() {
     println!("else if 2");
} else if bla4() {
     println!("else if 3");
} else if bla5() {
      //~^ ERROR: `if` expression with an `else if`, but without a final `else`
      println!("else if 4");
}
```

Before this fix, four duplicate lints are emitted for this `if-else-if...` chain.  

Also added a passing case four chain test, and fail/pass cases which contain suppressed [`collapsible_else_if`] blocks.